src/engine/thread.js

```diff
@@ -254,34 +280,41 @@ class Thread {
      * @param {string} blockId Block ID to push to stack.
      */
     reuseStackForNextBlock (blockId) {
-        this.stack[this.stack.length - 1] = blockId;
-        this.stackFrames[this.stackFrames.length - 1].reuse();
+        this.pointer = blockId;
+        if (this.stackFrame.needsReset) this.stackFrame.reset();
     }
 
     /**
-     * Pop last block on the stack and its stack frame.
-     * @return {string} Block ID popped from the stack.
+     * Move the instruction pointer to the last value before this stack of
+     * blocks was pushed and executed.
+     * @return {?string} Block ID popped from the stack.
      */
     popStack () {
-        _StackFrame.release(this.stackFrames.pop());
-        return this.stack.pop();
+        const lastPointer = this.pointer;
+        this.pointer = this.stack.pop() || null;
+        _StackFrame.release(this.stackFrame);
+        this.stackFrame = this.stackFrames.pop() || null;
+        return lastPointer;
     }
 
     /**
-     * Pop back down the stack frame until we hit a procedure call or the stack frame is emptied
+     * Move the instruction pointer to the last procedure call block and resume
+     * execution there or to the end of this thread and stop executing this
+     * thread.
      */
     stopThisScript () {
         let blockID = this.peekStack();
         while (blockID !== null) {
             const block = this.target.blocks.getBlock(blockID);
             if (typeof block !== 'undefined' && block.opcode === 'procedures_call') {
+                this.pushStack(null);
```

Comment:
Could you add a comment here about why this is significant?

Reply:
I added:

```js
// If we do not push this null, Sequencer will not step to the
// next block and will erroneously call the procedure a second
// time.
//
// By pushing null, Sequencer will treat it as the end of a
// substack, pop the stack and step to the next block.
```